### ISSUE TITLE
Incorrect docs for RET504

### ISSUE THREAD

**Issue by Apakottur** [charliermarsh/ruff#4240]
I am looking at the docs for [RET504 (unnecessary-assign)](https://beta.ruff.rs/docs/rules/unnecessary-assign/). The following code is given there as an example:

```python
# test.py
def foo():
    bar = 1
    # some code that not using `bar`
    print('test')
    return bar
```

When I run `ruff check --select ALL test.py` on this code I get:
```shell
warning: `one-blank-line-before-class` (D203) and `no-blank-line-before-class` (D211) are incompatible. Ignoring `one-blank-line-before-class`.
warning: `multi-line-summary-first-line` (D212) and `multi-line-summary-second-line` (D213) are incompatible. Ignoring `multi-line-summary-second-line`.
test.py:1:1: D100 Missing docstring in public module
test.py:1:5: ANN201 Missing return type annotation for public function `foo`
test.py:1:5: D103 Missing docstring in public function
test.py:4:5: T201 `print` found
Found 4 errors.
```

There is no RET504 error. Is the example incorrect or am I missing something?

**Comment by calumy** [comment]
@Apakottur, thanks for highlighting this. This is an example of where #4168 could be useful. 

I personally think that this is a false negative and should be flagged as a RET504 error, but it appears that the presence of the print statement stops this error from appearing. 

If the example was updated to:

```python
# test.py
def foo():
    bar = 1
    # some code that not using `bar`
    return bar
```

Then the error does appear:

```bash
> ruff test.py --select=RET
test.py:5:12: RET504 Unnecessary variable assignment before `return` statement
Found 1 error.
```

I see that there are a couple of solutions:
- Fix the rule so that this returns a RET504 error as the print statement does not use `bar`
- Update the docs if it is deemed that this is not a false negative. I'm happy to do this if needs be but would err on the side of this being a false negative

**Comment by Apakottur** [comment]
What about the following code:
```python
def func() -> int:
    match True:
        case True:
            x = 5
            return x
        case _:
            raise ValueError("Error")
```

I don't get `RET504` here as well. Is that the expected behavior?

**Comment by calumy** [comment]
I think it probably should return an error as it could be updated to:

```python
def func() -> int:
    match True:
        case True:
            return 5
        case _:
            raise ValueError("Error")
```

**Comment by dhruvmanila** [comment]
> I don't get `RET504` here as well. Is that the expected behavior?

No. After giving a quick look, it's occurring as the `Visitor` is not going through the `match` statement. A new issue would help keep track of this bug :)
